### DISCUSSION TITLE
Internal Dask usage

### DISCUSSION THREAD

**Discussion by saulshanabrook** [Quansight-Labs/metadsl#114] [Ideas]
It would be great to understand the Dask use cases for high level optimization well enough to create some preliminary integration that addresses them. See https://github.com/dask/dask/issues/4038

**Comment by dhirschfeld** [comment]
Potential overlap with PyTorch also - https://twitter.com/alexjc/status/1048962480343580672

**Comment by saulshanabrook** [comment]
@dhirschfeld thanks for that link!

Looks like lots of people are building their own lazy arrays for a particular combination of problem domain and backend. Question is, can we do something here that is more general purpose enough to be useful but not too general to be over complicated and hard to optimize.

**Comment by dhirschfeld** [comment]
And obviously https://github.com/mrocklin/symbolic-array/issues/1 (which you're aware of).

Also, there's the Gandiva Expression Compiler for Arrow
https://www.dremio.com/announcing-gandiva-initiative-for-apache-arrow/

Plus `xtensor` - https://xtensor.readthedocs.io/en/latest/expression.html#expressions-and-lazy-evaluation

...it seems like an idea whose time has come, but it seems that everyone is going their own way. Maybe there are good reasons for that, IDK, but as a user I'd much rather one good implementation than 10 half-baked ones.

With so many different initiatives I'm worried that none will gain enough traction to be sustainable into the future and that they'll go the way of `blaze` :(

Still, I'm very excited about the prospects! Optimisation of lazy expression graphs was one of the concepts which excited me about `blaze`. Whilst an experienced dev can write efficient (and parallel) code it often loses a lot of the clarity (and hence maintainability) of the original algorithm. Lazy experession graphs, optimised by the computer, promised the ability for less experienced devs to implement an algorithm naively (but clearly) and still achieve good performance (and parallelism with e.g. `dask`)

**Comment by saulshanabrook** [comment]
Yeah xtensor seems pretty nice! I still don't fully grok it, but from talking to a few of the authors I think they use a lot of C++ templating to help them do the compilation.

Even in scientific Python land, the idea seems to have caught on a while ago with [Theano](http://deeplearning.net/software/theano/introduction.html).

> Theano is a Python library that lets you to define, optimize, and evaluate mathematical expressions, especially ones with multi-dimensional arrays (numpy.ndarray). Using Theano it is possible to attain speeds rivaling hand-crafted C implementations for problems involving large amounts of data. It can also surpass C on a CPU by many orders of magnitude by taking advantage of recent GPUs.
> 
> Theano combines aspects of a computer algebra system (CAS) with aspects of an optimizing compiler. It can also generate customized C code for many mathematical operations. This combination of CAS with optimizing compilation is particularly useful for tasks in which complicated mathematical expressions are evaluated repeatedly and evaluation speed is critical. For situations where many different expressions are each evaluated once Theano can minimize the amount of compilation/analysis overhead, but still provide symbolic features such as automatic differentiation.

And has seen success with the influx of newer define-then-run deep learning frameworks (tensorflow, etc). 

---

So why should this project exist, given that lazy array libraries already exist?

To make it easier for end users to keep their same code while switching backends and to make it easier for backend developers to get new users without creating a new API.

The premise is that many of the different lazy array computation projects end up at some point or another in their compilation step end up with a similar API/algebra. Therefore, at that point, we are abstracted away from both the frontend UX and the backend, so developers of either can target this point and share each others work.

The opening up of NumPy's API for other backends is a way to doing that where the middle group is NumPy's API (see slide from NumPy summit @ SciPy 2018):

![mvimg_20180712_131347](https://user-images.githubusercontent.com/1186124/47864092-314e8b80-ddcf-11e8-9a2e-fb1e382b999f.jpg)


Uarray extends that high level API by allowing backends to say "Hey I know how to do indexing and scalar operations and looping and etc, now target me as your backend!"

This can work by starting with a user interface of existing NumPy API and translating, or it can start from any other UX, which all just build up array expression trees. So the aim is to let users create any new operators they like, define translations for those to more basic operators or directly to backends, and provide some on-ramps to this process through familiar APIs like that of NumPy.

---

So looking into Dask's implementation of high level graphs (https://github.com/dask/dask/pull/4092) might be a good user test to see uarray can help perform that function.

**Comment by saulshanabrook** [comment]
Links to relevant dask issues:

* Dask (source: conversation w/ @mrocklin)
  * Optimize expressions before turning to graph (reduces size of graph, removes intermediaries) https://github.com/dask/dask/issues/2538 https://github.com/dask/dask/issues/3514 https://github.com/dask/dask/pull/3549 https://github.com/dask/dask/issues/746 https://github.com/dask/dask/pull/2608
  * More intelligent automatic chunking optimizations based on operations (https://github.com/dask/dask/issues/3864) (UK Met office might be interested in this)

**Comment by jakirkham** [comment]
If you were able to propose something reasonable for issue ( https://github.com/dask/dask/issues/746 ) or cases like that, there are many people doing image processing with dask and/or xarray that would really appreciate that improvement. Given your talk at PyData DC the other day, suspect that you are already able to solve this class of problems nicely. So adding an example for this class of problems, would be a really good place to start (assuming there isn't one already).

cc @d-v-b @shoyer

**Comment by dhirschfeld** [comment]
It looks like work has started in the arrow repo on an expressin graph implementation:
https://github.com/apache/arrow/pull/3820

**Comment by wesm** [comment]
As context we are in the early stages of building a query engine runtime in the style of modern analytic databases ("Volcano-style") -- we need a well-typed query AST that can be lowered to a physical execution plan (containing scans, projections, joins, aggregations/hash aggregations, and other database-y type things). Apache Impala is an example of such a system but more complex (because it's a distributed system -- we are only really interested in single node multicore right now) https://github.com/apache/impala/tree/master/be/src/exec
